added spaces documentation (#12404)
RasaHQ/rasa: docs/docs/spaces.mdx

Reviewer: Maybe we can add the extra section illustrating the `--space` argument (to train separate space)?

In addition it might be useful to have a showcase of financial bot (illustrating the difference between separate section training vs full bot training)

Author: Good idea, I am not yet familiar with the syntax we settled on in the end

Reviewer 2: the test and shell commands don't require this as they load a trained model

Reviewer 3: thanks for flagging this Greg. We will think about this for the GA

---
id: spaces
sidebar_label: Spaces

> title: Spaces

Author: I made another run and capitalized the word when it refers to the Spaces feature. However, in sentences like "you can see three spaces" I left it lower case.

> title: Spaces

Reviewer 3: totally makes sense :)

description: Learn about Spaces for Rasa.
abstract: Spaces are a way to modularize your assistant.
---

import RasaProLabel from "@theme/RasaProLabel";
import RasaProBanner from "@theme/RasaProBanner";
import useBaseUrl from '@docusaurus/useBaseUrl';

<RasaProLabel />

<RasaProBanner />

:::caution New in 3.6
This feature is experimental.
We introduce experimental features to get feedback from our community, so we encourage you to try it out!
However, the functionality might be changed or removed in the future.
If you have feedback (positive or negative) please share it with us on the [Rasa Forum](https://forum.rasa.com).

> :::

Reviewer 4: we'll go with "alpha"

to add some specific space-aware components to your NLU pipeline. The following is the
`config.yml` of the financial spaces bot linked above:

```yaml
# https://rasa.com/docs/rasa/model-configuration/
recipe: default.v1

# Configuration for Rasa NLU.
# https://rasa.com/docs/rasa/nlu/components/
language: en

spaces:
  - name: main
    domain: main/domain.yml
    nlu: main/nlu/training_data.yml
    nlu_test: main/nlu/test_data.yml
    rules: main/rules.yml
  - name: transfer_money
    domain: transfer_money/domain.yml
    nlu: transfer_money/nlu/training_data.yml
    nlu_test: transfer_money/nlu/test_data.yml
    rules: transfer_money/rules.yml
    entry_intents:
      - transfer_money
  - name: investment
    domain: investment/domain.yml
    nlu: investment/nlu/training_data.yml
    nlu_test: investment/nlu/test_data.yml
    rules: investment/rules.yml
    entry_intents:
      - buy_stock
  - name: pay_cc
    domain: pay_cc/domain.yml
    nlu: pay_cc/nlu/training_data.yml
    nlu_test: pay_cc/nlu/test_data.yml
    rules: pay_cc/rules.yml
    entry_intents:
      - pay_cc

importers:
  - name: "rasa_plus.spaces.space_data_importer.SpaceDataImporter"

>     temporary_working_directory: spaces

Reviewer 3: @twerkmeister there's this minor comment that you missed I think

pipeline:
  - name: WhitespaceTokenizer
  - name: RegexFeaturizer
  - name: LexicalSyntacticFeaturizer
  - name: CountVectorsFeaturizer
  - name: CountVectorsFeaturizer
    analyzer: char_wb
    min_ngram: 1
    max_ngram: 4
  - name: rasa_plus.spaces.components.spaces_crf_entity_extractor.SpacesCRFEntityExtractor
  - name: EntitySynonymMapper
  - name: DIETClassifier
    epochs: 100
    ranking_length: 0
    entity_recognition: false
    BILOU_flag: false
  - name: rasa_plus.spaces.components.filter_and_rerank.FilterAndRerank
  - name: FallbackClassifier
    threshold: 0.3
    ambiguity_threshold: 0.1

policies:
  - name: RulePolicy

```

In the above config file, you can see the spaces definition with the main space, which
is a designated name for a space for shared primitives, and the three specialized
spaces for transferring money, investment, and credit card payments. Each path value

> can either point to a single file or a directory.

Author: is it required?

> can either point to a single file or a directory.

Reviewer 3: I assume it is (we keep mentioning it like it is). What happens if people don't have a main space? Does the feature crash?

> can either point to a single file or a directory.

Author: It is not required. It might also be empty. It's only for sharing things. If nothing is shared, then it is not needed :)

Second, you can also see the importer
`rasa_plus.spaces.space_data_importer.SpaceDataImporter` being used. This importer
reads the above spaces definition and assembles the bot. The results can be seen in the
temporary working directory.

Third, you can see the specific NLU components that are necessary to make spaces work.
The most important is the `FilterAndRerank` component which is central to make spaces
work by post-processing intent rankings and entity extractions. If you also need entity
extraction you need to use the `SpacesCRFEntityExtractor` to make that work with spaces.

Finally, you can see that we are only using the `RulePolicy` here.

After you have created your spaces and made adjustments to the `config.yml`, you can
train your new assistant using `rasa train -c config.yml` and start it afterward with
`rasa shell`.

## How do spaces work?
In the following we'll look at how things work under the hood in spaces to give you a
better understanding of what is happening in case you should need it.

One of the most aspects is that spaces form a hierarchy with two layers. At the top
is the main space which contains shared primitives that should be usable by all spaces.

> The main space is always active. Anything that is defined in a subspace, can

Reviewer 3: `subspace`

I think we need to consistent in the wording here. We've used `specialized space` internally. @rosanardila can you make a decision on that specific bit?

only be used by that space and not by other subspaces or the main space.

<img alt="two layer space hierarchy" src={useBaseUrl("/img/spaces_hierarchy.png")} />

### What happens during the assembly?

The most important step during the assembly is the prefixing. During this step every
intent, entity, slot, action, form, utterance that is defined in a space's domain file

> is prefixed (infixed for utterances) with this space's name. For example, an intent

Reviewer 3: I'm not sure what "(infixed for utterances)" means

Reviewer: Agree, prefixing of utterances should be shown in detail (to avoid confusion during log reading, for example)

`ask_transfer_charge` in the `transfer_money` domain would become
`transfer_money.ask_transfer_charge` and every reference of this intent would be
adjusted. The final assistant then works on the prefixed data.

An exception to this is the main space. Anything in the main space and all its
primitives that are used in other spaces, will not be prefixed.

> Another exception are rules. They don't have a name that can be prefixed. Instead,

Reviewer 3: I'd add a hyperlink to "rules"

we add a condition to each rule that it is only applicable while it's space is active.

> ### How is space activation tracked?

Reviewer 3: I think we can rework this section to "How does space activation work?" and mention entry intents + the bit about slots below

> Space activation is currently tracked through boolean slots that are automatically

Reviewer: Are those slots are accessible to the users? If not, this might be too low level implementation details, not relevant to the user. 

Instead, we might describe overall logic of space activation to give the idea of what's going on

generated during assembly for each space.

### How does filter and rerank work?

> The filter and rerank component post-processes the intent ranking of the intent

Reviewer 3: maybe not for the alpha, but ultimately we should describe this component with the others in [this page](https://rasa.com/docs/rasa/components)

> The filter and rerank component post-processes the intent ranking of the intent
> classification components. It accesses the tracker and checks which space are active

Reviewer 3: ```suggestion
The filter and rerank component post-processes the intent ranking of the [intent
classification components](./components.mdx#intent-classifiers). It accesses the tracker and checks which space are active
```

or would be activated, in case an entry intent is at the top.

It then removes any intents from the ranking that are not possible. Further it also
removes any entities that are not possible given the space activation status or
about to be predicted entry intents.

### How does entity recognition work differently?

Usually, entity recognizers in Rasa only return a single label per token in a message.
Thus, there is no ranking, that could be post-processed as in the case of intents. We

> have built our `SpacesCRFEntityExtractor` in a way that it creates multiple extractors.

Reviewer 3: maybe not for the alpha, but ultimately we should describe this component with the others in [this page](https://rasa.com/docs/rasa/components)

> One for each space. Now, during the post-processing step, we can filter out the

Reviewer: I think this short sentence "One for each space" could be joined to the previous one

extractor of the spaces that are not activated.

> ### Custom Actions

Reviewer 3: > Custom actions from the main spaces are not inherited by specialised spaces;

should we also mention the limitation listed [here](https://www.notion.so/rasa/Spaces-Implementation-Proposal-updated-Q2-2023-1a834e7203cb47cdabfb46049a0c983d?pvs=4#84b982bf81a14d7690a09a0e3df348c2)?

Custom actions work as before. However, the tracker, the domain, and slots
will be stripped of any information from other spaces before being handed to your
action. Additionally, every event such as slot sets, will be prefixed after they are

> returned by your action. All of this ensures that from the view point of your custom

Reviewer 3: ```suggestion
returned by your action. All of this ensures that from the viewpoint of your custom
```

action, you don't need to worry about the other spaces and accidentally leaking

> or altering information. It is as if the other spaces and prefixing does not exist.

Reviewer 3: ```suggestion
or altering information. This warrants isolation between your Spaces.
```

### Response Selection

Response selection works as before. The only small difference is that in your

> `config.yaml` you'll need to specify the retrieval intent including it's final prefix.

Reviewer 3: ```suggestion
`config.yml` you'll need to specify the retrieval intent including its final prefix.
```

If your retrieval intent is `investment_faq` in the `investment` space, then in your
config you'll need to set `investment.investment_faq` as the retrieval intent. If the
retrieval intent belongs to the main space, no prefix is added. That's why
, in this case, during the definition of the retrieval intent, no prefixing is necessary.

> 

Reviewer 3: can we also mention synonyms and lookup tables, together with limitations if any?

Author: there were non for the synonyms, right? Are there any for the lookup tables @Shmuma ?

Reviewer 3: but at least mentioning that the feature works with both of them (no limitation for lookup tables; some limitations for synonyms) would be good. I've already described those in [this section](https://www.notion.so/rasa/Spaces-Implementation-Proposal-updated-Q2-2023-1a834e7203cb47cdabfb46049a0c983d?pvs=4#12c7d0c51a3d49f6ad9eb49c40e4f8eb)

---
id: spaces
sidebar_label: Spaces
title: Spaces
description: Learn about Spaces for Rasa.
abstract: Spaces are a way to modularize your assistant that increases isolation between
subparts, and classification performance for intents and entities that are only
relevant in specific contexts.
---

import RasaProLabel from "@theme/RasaProLabel";
import RasaProBanner from "@theme/RasaProBanner";
import useBaseUrl from '@docusaurus/useBaseUrl';

<RasaProLabel />

<RasaProBanner />

:::caution New in 3.6.0a1

> This feature is experimental.

Reviewer 4: This is an alpha release

> We introduce experimental features to get feedback from our community, so we encourage you to try it out!

Reviewer 4: This alpha release will allow us to gather feedback from our users and integrate it into future releases. We encourage you to try this out!

> However, the functionality might be changed or removed in the future.

Reviewer 4: The functionality of this alpha release might change in the future.

every of the existing features of rasa. We made it work with almost all of them, but
there are some exceptions:

* if you have created extensive customization and introduced your own data hierarchies
or post-processing steps, these might not be compatible with Spaces.
* Spaces currently do not support stories. This is a big limitations for many
existing bots. However, because stories and their existing policies work with a fixed
event horizon (`max_history`), they are at this point not compatible with the idea of
being able to describe encapsulated units of logic well.
* Entity extraction is currently limited to an adapted version of
the `CRFEntityExtractor` which works with Spaces boundaries. You can find more on this
in the following sections.

## An example Spaces bot
We have created a bot using Spaces for you to look at, learn from and experiment with.
It features three different spaces in the financial domain. [Here is the repository](https://github.com/RasaHQ/financial-spaces-bot).

## How to use Spaces

Spaces are defined by separate sets of domain, nlu, and rule files. These separate
sets are then assembled to create a unified assistant. The assembly plan needs to be

> added to the `config.yml` using the newly introduces `spaces` key. Second, you need

Reviewer 5: Shouldn't this be in `domain.yml` since it's focused on the training data primitives that are also in `domain.yml` and not the pipeline and polices?

Reviewer 3: ✅ replied on Slack [here](https://rasa-hq.slack.com/archives/C01TD639GMQ/p1684480173880869?thread_ts=1684337607.585739&cid=C01TD639GMQ) and [here](https://rasa-hq.slack.com/archives/C01TD639GMQ/p1684482353729519?thread_ts=1684337607.585739&cid=C01TD639GMQ). This is something we can revisit with customer feedback for the GA

is a designated name for a space for shared primitives, and the three
subspaces for transferring money, investment, and credit card payments. The main space
is not required, but the only way to share primitives between spaces. Each path value
can either point to a single file or a directory.

Second, you can also see the importer
`rasa_plus.spaces.space_data_importer.SpaceDataImporter` being used. This importer
reads the above Spaces definition and assembles the bot. The results can be seen in the
configurable temporary working directory.

Third, you can see the specific NLU components that are necessary to make Spaces work.
The most important is the `FilterAndRerank` component which is central to make Spaces
work by post-processing intent rankings and entity extractions. If you also need entity
extraction you need to use the `SpacesCRFEntityExtractor` to make that work with Spaces.

Finally, you can see that we are only using the `RulePolicy` here.

After you have created your spaces and made adjustments to the `config.yml`, you can
train your new assistant using `rasa train -c config.yml` and start it afterward with
`rasa shell`.

### Training only a specific subspace

> We have added `--space` argument to the rasa train command to give you the option to

Reviewer 5: How does this affect use of the `--data` and `--domain` cli options. If those options are used and there's a spaces config are the options assumed be used for the main space or are these options ignored when a spaces config is in use?

Reviewer 3: I replied to you on Notion:
> Indeed, the rasa train command only support either --space or --domain or --data now. Not a combination of those.

I think it's fine to leave that out of the docs for the alpha -- the CLI will yield a descriptive error if a user attempts to use them in combination
